File: pippip-native-ios/API/AuthenticationResponse.m
//
//  AuthenticationResponse.m
//  pippip-native-ios
//
//  Created by Steve Brenneis on 12/4/17.
//  Copyright © 2017 seComm. All rights reserved.
//

#import "AuthenticationResponse.h"
#import "pippip_native_ios-Swift.h"
#import "CKRSACodec.h"

@interface AuthenticationResponse () {

    SessionState *sessionState;

}
@end

@implementation AuthenticationResponse

- (instancetype)init {
    self = [super init];
    
    sessionState = [[SessionState alloc] init];
    return self;
    
}

- (BOOL)processResponse:(NSDictionary *)response errorDelegate:(id<ErrorDelegate>)errorDelegate {
    
    NSString *dataStr = [response objectForKey:@"data"];
    NSString *errorStr = [response objectForKey:@"error"];
    if (errorStr != nil) {
        [errorDelegate responseError:errorStr];
        return NO;
    }
    if (dataStr == nil) {
        [errorDelegate responseError:@"Invalid server response"];
        return NO;
    }
    else {
        NSData *data = [[NSData alloc] initWithBase64EncodedString:dataStr options:0];
        if (data == nil) {
            [errorDelegate responseError:@"Server encoding error"];
            return NO;
        }
        else {
            NSError *error;
            CKRSACodec *codec = [[CKRSACodec alloc] initWithData:data];
            [codec decrypt:sessionState.userPrivateKey withError:&error];
            if (error != nil) {
                [errorDelegate responseError:[error localizedDescription]];
                return NO;
            }
            sessionState.serverAuthRandom = [codec getBlock];
            return YES;
        }
    }
    
}

@end

File: pippip-native-ios/API/AuthenticationResponse.swift
//
//  AuthenticationResponse.swift
//  pippip-native-ios
//
//  Created by Steve Brenneis on 5/29/18.
//  Copyright © 2018 seComm. All rights reserved.
//

import UIKit
import ObjectMapper
import CocoaLumberjack

class AuthenticationResponse: APIResponseProtocol {
    
    var error: String?
    var sessionId: Int32?
    var authToken: Int64?
    var data: String?
    var postId: Int = 0

    var alertPresenter = AlertPresenter()
    var sessionState = SessionState()

    required init?(map: Map) {
        if map.JSON["error"] == nil {
            if map.JSON["sessionId"] == nil {
                return nil
            }
            if map.JSON["data"] == nil {
                return nil
            }
        }
    }

    func mapping(map: Map) {

        sessionId <- map["sessionId"]
        error <- map["error"]
        data <- map["data"]

    }
    
    func processResponse() throws {

        if let responseError = error {
            throw APIResponseError.serverResponseError(error: responseError)
        }

        if let encoded = Data(base64Encoded: data!) {
            let codec = CKRSACodec(data: encoded)
            try codec.decrypt(sessionState.userPrivateKey!)
            sessionState.serverAuthRandom = codec.getBlock()
        }
        else {
            throw APIResponseError.invalidServerResponse
        }

    }
    
}
